R/insert-ui.R

```diff
@@ -0,0 +1,194 @@
+#' wrapped in an extra div before making its way into the DOM. This
+#' does not affect what you mean to do, and it makes it easier to remove
+#' the whole UI object using \code{\link{removeUI}} (if you wish to do
+#' so, of course).
+#'
+#' @param selector A string that is accepted by jQuery's selector (i.e. the
+#' string \code{s} to be placed in a \code{$(s)} jQuery call). This selector
+#' will determine the element(s) relative to which you want to insert your
+#' UI object.
+#'
+#' @param multiple In case your selector matches more than one element,
+#' \code{multiple} determines whether Shiny should insert the UI object
+#' relative to all matched elements or just relative to the first
+#' matched element (default).
+#'
+#' @param where Where your UI object should go relative to the selector:
+#' \describe{
+#'   \item{\code{beforeBegin}}{Before the selector element itself}
+#'   \item{\code{afterBegin}}{Just inside the selector element, before its
+#'   first child.}
+#'   \item{\code{beforeEnd}}{Just inside the selector element, after its
+#'   last child}
+#'   \item{\code{afterEnd}}{After the selector element itself (default)}
+#' }
+#' Adapted from
+#' \href{https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentHTML}{here}.
+#'
+#' @param ui The UI object you want to insert. This can be anything that
+#' you usually put inside your apps's  \code{ui} function.
+#'
+#' @param immediate Whether the UI object should be immediately inserted into
+#' the app when you call \code{insertUI}, or whether Shiny should wait for the
+#' next time your app is updated for some other reason (default).
+#'
+#' @param container A function to generate an HTML element to contain the UI
+#' object.
+#'
+#' @param inline Use an inline (\code{span()}) or block container (\code{div()},
+#' default) for the output.
+#'
+#' @param session The shiny session within which to call \code{insertUI}.
+#'
+#' @seealso \code{\link{removeUI}}
+#'
+#' @examples
+#' ## Only run this example in interactive R sessions
+#' if (interactive()) {
+#' # Define UI
+#' ui <- fluidPage(
+#'   actionButton("add", "Add UI")
+#' )
+#'
+#' # Server logic
+#' server <- function(input, output, session) {
+#'   observeEvent(input$add, {
+#'     insertUI(
+#'       selector = "#add",
+#'       ui = textInput(paste0("txt", input$add),
+#'                      "Insert some text")
+#'     )
+#'   })
+#' }
+#'
+#' # Complete app with UI and server components
+#' shinyApp(ui, server)
+#' }
+#'
+#' @export
+insertUI <- function(selector,
+  multiple = FALSE,
+  where = c("beforeBegin", "afterBegin", "beforeEnd", "afterEnd"),
+  ui,
+  immediate = FALSE,
+  container = if (inline) "span" else "div",
+  inline = FALSE,
+  session = getDefaultReactiveDomain()) {
+
+  force(selector)
+  force(ui)
+  force(session)
+  if (missing(where)) where <- "afterEnd"
+  where <- match.arg(where)
+
+  sendInsertUI <- function() {
+    session$sendCustomMessage("shiny-insert-ui",
```

Comment:
Sorry, I think we're supposed to use `session$sendMessage` instead of `sendCustomMessage` for Shiny-internal features. I've been in the bad habit of using `sendCustomMessage` in the past but @wch fixed all of them I think. When you use `sendMessage` the way you receive messages on the client side is a little different, you use `addMessageHandler` instead of `exports.addCustomMessageHandler`.